### ISSUE TITLE
Feature Request: Warn when mass rebuild PR has wrong base

### ISSUE THREAD

**Issue by matthewbauer** [NixOS/ofborg#66]
This is a small feature request I have that would be helpful for PRs.

Basically, when either 'rebuild-darwin: 501+' or 'rebuild-linux: 501+' is found, check what the base branch is. If it's based off of 'master', post a warning message from @GrahamcOfBorg explaining that it should be based off of 'staging' instead (except in rare circumstances).

**Comment by grahamc** [comment]
I think this is a really good idea, but am afraid of adding _more_ comments to pull requests :(

**Comment by matthewbauer** [comment]
I hope it doesn't happen too much! I guess there might be other ways to warn with GitHub APIs? Some other options:

- regular comment (original)
- PR review comment
- additional status

**Comment by 7c6f434c** [comment]
I think it is OK to add more comments for the situations that should be rare and special.

At the same times something about comment coalescing for the common case would be nice, yes…

**Comment by grahamc** [comment]
What if instead of a comment it failed the PR's status?

if master & 501+ => fail: please move to staging

cc @vcunat

**Comment by 7c6f434c** [comment]
Not sure this is a good idea… Does this mean that a `rust-unstable` update must go to `staging` once there are enough `cargo` packages?

**Comment by grahamc** [comment]
I suppose yes,

> Does this mean that a ≪compiler≫ update must go to staging once there are enough dependent packages?

if the goal is to avoid serious inconvenience for contributors, and there are hundreds of packages depending upon ≪compiler≫ then yes... right?

**Comment by shlevy** [comment]
I think the usage percentage might matter... But as a starting point let's try it as pure numbers.

**Comment by grahamc** [comment]
Indeed --  `501+` -> staging has always sort of been loose policy. Failing the PR would make it less soft policy and more hard policy.

**Comment by 7c6f434c** [comment]
>> Does this mean that a ≪compiler≫ update must go to staging once there are enough dependent packages?
>
>if the goal is to avoid serious inconvenience for contributors, and there are hundreds of packages depending upon ≪compiler≫ then yes... right?

Well, when there are hundreds of small autoconverted packages, the inconvenience of going through staging is larger than rebuilding the twenty you actually need now…

**Comment by grahamc** [comment]
Yeah, I don't know what to say other than rules are imperfect.

**Comment by grahamc** [comment]
It would be very inconvenient to be the person who depends on those 501+ rust packages and someone pushes a 501+ rust rebuild to master. At the end of the day is what is the 501+ rule for and why do we have staging, and how do those intentions translate in to the goals of the check.

**Comment by grahamc** [comment]
I mean, a separate conversation is 501+ even the line we want to draw? It was just sort of picked out of a hat when I started tagging.

**Comment by 7c6f434c** [comment]
I think 501 Python packages are less annoying to rebuild than a single Chromium. 

Also, I would expect that unstable rustPackages or lispPackages or pythonPackages have a different typical usage — do we have a person who depends on 500 rust packages?

**Comment by grahamc** [comment]
I don't think that level of detail in the rules is a feasible thing right now.

**Comment by grahamc** [comment]
If knowing 500 things rebuilds is not useful, why have the tag at all?

Also, for example the fail could be done only if stdenv is changed.

**Comment by shlevy** [comment]
@grahamc For what it's worth, I'd love to get to the point where we can ditch staging altogether and have the `-unstable` channels be actually reliable ways to stay up-to-date without rebuilding too much.Maybe eventually we can have ofborg block a PR if the number of packages is too high, but have a command to do a build and populate the cache and then it can be merged.

**Comment by shlevy** [comment]
I think it's a useful hint, but it's not definite. IMO we should just give it a shot to have 500+ cause a failure and just let trusted people bypass it for now. If it turns out to be obnoxious we can iterate.

**Comment by grahamc** [comment]
My only hesitation is I'm afraid of making PR checks fail for reasons that can be ignored. As it stands, there is no valid reason to merge a PR when its status is red.

**Comment by 7c6f434c** [comment]
I support failure on `stdenv` change in `master`. A hand-picked set of extra critical libraries is also reasonable.

I think `501+` label is very useful to compare expectations and reality. I.e. «of course 500 packages changed, it's a change in Python» vs. «wait what why, what did I miss?»

I agree that going the TravisCI-for-Nixpkgs way where «you won't believe me, but the Travis failure is actually relevant» happens, is not attractive.

**Comment by LnL7** [comment]
I agree with @grahamc, that we should avoid introducing failures that are "ok" to be ignored.

I think 500 is probably a bit of a low threshold to disallow a merge. Given that master progresses pretty quickly it's probably impossible for a normal staging merge not to cross that for example. Not sure what a reasonable number is but failing on a stdenv rebuild definitively sounds reasonable to me.

**Comment by Profpatsch** [comment]
As another input, updating a big npm-based project’s dependencies is probably also hit that 500 derivation policy. 200+ fixed input derivations (the sources) plus 200+ package derivations plus one `node_modules` derivation for every non-leaf dependency.

**Comment by grahamc** [comment]
ofborg counts based on changed named attributes, not based on total changed derivations,h so it should count an npm-based project as one.

**Comment by vcunat** [comment]
Version bumps of the default kernel are hitting 501+ IIRC.

**Comment by grahamc** [comment]
Maybe you can offer some insight in to what you would find useful, and possibly better package-rebuild buckets?

**Comment by vcunat** [comment]
Maybe one more level for Linux, say 500&ndash;2000?  (The <10 level might get ditched, too.)  EDIT: it might be a gray area where some cases are more suitable for master and some for staging.

I'm wary of shifts, as I don't know what we would do with already labeled PRs, though it might not be too difficult to just run eval on all of them again.

**Comment by grahamc** [comment]
Sounds good.

 - Would it be proper to fail a rebuild-2000+ PR to master, saying it should be redirected to staging?
 - Would it be proper to fail a rebuild-stdenv PR to master, saying it should be redirected to staging?

I can do the shift in such a way that the old 500+ tag will get removed upon re-eval. Thus, on PRs we're not sure on, we can just trigger a re-eval.

**Comment by vcunat** [comment]
It seems difficult to make these _hard_ rules, as sometimes changes have a separate jobset so binaries are ready at that moment.  I think they usually aren't merged via PRs, though.

I'm still not really sure how to do such workflows, as such branches still often interfere with effectivity of staging (in terms of minimizing total rebuilds), but it seemed typically only a minor issue so far.

**Comment by grahamc** [comment]
Makes sense, in this case I think the proper solution is probably not to fail the build at all, and fall back to a soft recommendation of using staging via a comment or something.

**Comment by 7c6f434c** [comment]
Re: PRs vs. direct push — maybe I should make it explicit that I argue from the assumption that we do not want to do a borg policy hard turn between now and the beginning of «maybe no direct push to master» talk.

**Comment by 7c6f434c** [comment]
Another thing that can change the balance is a notable bugfix (a security issue, a data integrity risk)…

**Comment by FRidh** [comment]
I would like to see another tag for `500 - 5000` *and* the number of rebuilds posted *somewhere*. In my opinion we should forbid merging to master at  `5000+`. Having that big changes on master slows down `staging-next` (and all other jobs), and basically means an extra day of delay because `staging-next` will need to rebuild that set of changes *again* before it is merged into `master`.

At `500 - 5000` it should be discouraged. I think @r-ryantm already changes branch to staging at `500+`. 

Edit: @LnL7 pointed out that, when clicking on the eval check, you're presented with the whole list of items to rebuild.

**Comment by ryantm** [comment]
> I think @r-ryantm already changes branch to staging at 500+.

I manually change @r-ryantm's PRs when I see the "500+" tag. I don't always get to it though.

**Comment by grahamc** [comment]
What if there was a specific canary attribute which was the rule for requiring it go to staging?

**Comment by ryantm** [comment]
If OfBorg removed it after it switched branches, this would make it very easy to search for the PRs that need to switch to staging.

**Comment by grahamc** [comment]
Definitely would be part of the feature I think.

**Comment by Ericson2314** [comment]
If we had separate statuses one could be like ofborg-optional vs ofborg-mandatory.

**Comment by FRidh** [comment]
Note that this last week there have been 2 mass-rebuilds on master.
